### ISSUE TITLE
Tx submission protocol: the speed of requests

### ISSUE THREAD

**Issue by denisshevchenko** [input-output-hk/ouroboros-network#1486]
Tx generator (as a part of `cardano-cli`) interacts with `ouroboros-network` (https://github.com/input-output-hk/cardano-node/blob/master/cardano-node/src/Cardano/CLI/Tx/BenchmarkingNodeToNode.hs#L71-L78), and this is what I see in generator's log (it's like a protocol dialog):

Hi guys! I have a question about tx submission protocol. That's what I see in tx generator's log. AFAIU, it's like a dialog:

1. Give me some txs:

```
{"at":"2020-01-20T14:33:16.74Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Recv ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgRequestTxIds TokNonBlocking 0 3"...
```

2. Ok, there's 3 txs ids for you:

```
{"at":"2020-01-20T14:33:16.77Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Send ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgReplyTxIds (NonBlockingReply [(txid: txid:f4d345ef,309),(txid: txid:18584784,309),(txid: txid:b2612f1c,309)])"...
```

3. Now give me two from these txs:

```
{"at":"2020-01-20T14:33:16.80Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Recv ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgRequestTxs [txid: txid:18584784,txid: txid:b2612f1c]"...
```

4. Take them:

```
{"at":"2020-01-20T14:33:16.80Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Send ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgReplyTxs [tx: Tx 18584784ff5409ea with inputs [TxInUtxo 0027bde8 #4], outputs: [TxOut 100000000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx, TxOut 201488999000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx], attributes: Attributes { data: (), remain: <0 bytes> }\nwitnesses: [\n    VKWitness: key = pub:08eead5d, key hash = 0738f782, sig = <signature>\n],tx: Tx b2612f1c8af305e2 with inputs [TxInUtxo 0027bde8 #5], outputs: [TxOut 100000000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx, TxOut 201488999000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx], attributes: Attributes { data: (), remain: <0 bytes> }\nwitnesses: [\n    VKWitness: key = pub:08eead5d, key hash = 0738f782, sig = <signature>\n]]","pid":"7741","loc":null,"host":"","sev":"Debug","thread":"1067"}
```

5. Now give me third tx as well:

```
{"at":"2020-01-20T14:33:16.83Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Recv ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgRequestTxs [txid: txid:f4d345ef]"...
```

6. Take it:

```
{"at":"2020-01-20T14:33:16.83Z","env":"fromList [(\"value\",String \"Release-1.0.0\")]:0.1.10.1","ns":["cardano","cli","generate-txs","submit"],"data":{},"app":[],"msg":"Send ConnectionId {localAddress = 172.31.45.235:47440, remoteAddress = 18.196.133.111:3001} MsgReplyTxs [tx: Tx f4d345ef05e04551 with inputs [TxInUtxo 0027bde8 #3], outputs: [TxOut 100000000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx, TxOut 201488999000 lovelace -> 2cWKMJemoBaiYicucs8ghsGVxxA6PMpVHPvky9Dc9HpaYQqH1mr2xerFBxK1kBZK5cwdx], attributes: Attributes { data: (), remain: <0 bytes> }\nwitnesses: [\n    VKWitness: key = pub:08eead5d, key hash = 0738f782, sig = <signature>\n]]","pid":"7741","loc":null,"host":"","sev":"Debug","thread":"1067"}
```

So:

`14:33:16.74` -> give me some txs ids
`14:33:16.77` -> take 3 ids
`14:33:16.80` -> give me 2 from these txs
`14:33:16.80` -> take them
`14:33:16.83` -> give me third tx
`14:33:16.83` -> take it

`ouroboros-network` asked 2 txs and then 1 tx, but not _all 3 txs at once_. It's a waste of time, isn't it? AFAIK, this because these hardcoded values: https://github.com/input-output-hk/ouroboros-network/blob/master/ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs#L128-L129.

There're two questions:

1. Why are these values (`maxTxIdsToRequest` and `maxTxToRequest`) hardcoded ones? And why `3` and `2`? Shouldn't they be configurable?
2. `cardano-cli` cannot just submit transactions, but only after `ouroboros-network` asks them (`MsgRequestTxs`). Does it mean that If these values (`maxTxIdsToRequest` and `maxTxToRequest`) would be increased - the speed of interaction between `ouroboros-network` and `cardano-cli` would be increased as well?

**Comment by dcoutts** [comment]
> Why are these values (maxTxIdsToRequest and maxTxToRequest) hardcoded ones? And why 3 and 2? Shouldn't they be configurable?

You link to this bit of code, where the comment answers your question.
```
    --TODO: replace these fixed limits by policies based on TxSizeInBytes
    -- and delta-Q and the bandwidth/delay product.
    -- These numbers are for demo purposes only, the throughput will be low.
    maxTxIdsToRequest = 3 :: Word16
    maxTxToRequest    = 2 :: Word16
```

**Comment by denisshevchenko** [comment]
Is it an answer? I see `TODO: replace...` So, it _should/will be_ replaced?

**Comment by karknu** [comment]
We increased the values connected to a similar deltaQ-Todo for blockfetch. This should be done here to.

**Comment by dcoutts** [comment]
So the real limit on pipelining is the `maxUnacked` these two just affect batching size.

The `maxUnacked` is currently hard coded to 100 in `Ouroboros.Consensus.Node`. These two batch sizes could quite reasonably be set to 25 and 10 respectively.

**Comment by denisshevchenko** [comment]
New results are (`--tps` is a generator's CLI argument):

`--tps 100` - max block size is `349KB`, real TPS is `71.86`
`--tps 250` - max block size is `537KB`, real TPS is `100.34`
`--tps 500` - max block size is `537KB`, real TPS is `125.00`

The reason is some hardcoded values in ouroboros-network, Luke helped me, so I increased them:

`mempoolCap`: `1000000` instead of `128000` (https://github.com/input-output-hk/ouroboros-network/compare/try-bigger-requests-sizes#diff-5bd30c844bbefdb3360cbe3077758ff9L282)
`maxTxIdsToRequest`: `250` instead of `3` (https://github.com/input-output-hk/ouroboros-network/compare/try-bigger-requests-sizes#diff-3aad85f6ecd6a0fb95bb1964916c7ff1L128)
`maxTxToRequest`: `100` instead of `2` (https://github.com/input-output-hk/ouroboros-network/compare/try-bigger-requests-sizes#diff-3aad85f6ecd6a0fb95bb1964916c7ff1L129)

In other words, now `ouroboros-network` asks much more txs (I see it in generator's log) and can receive much more txs at once.

**Comment by denisshevchenko** [comment]
`--tps 100`

```
 slot_no | slot_leader | epoch_no | block_no | txcount |  size  |        time         |  delta   
---------+-------------+----------+----------+---------+--------+---------------------+----------
         |           1 |          |          |     259 |      0 | 2020-01-23 05:24:02 | 
     272 |           3 |        0 |        4 |     107 | 393261 | 2020-01-23 06:54:42 | 01:30:40
     273 |           4 |        0 |        5 |      61 | 225049 | 2020-01-23 06:55:02 | 00:00:20
     274 |           5 |        0 |        6 |    1600 | 266552 | 2020-01-23 06:55:22 | 00:00:20
     275 |           3 |        0 |        7 |    2000 | 333208 | 2020-01-23 06:55:42 | 00:00:20
     276 |           4 |        0 |        8 |    1900 | 316532 | 2020-01-23 06:56:02 | 00:00:20
     277 |           5 |        0 |        9 |    2000 | 333188 | 2020-01-23 06:56:22 | 00:00:20
     278 |           3 |        0 |       10 |    2100 | 349860 | 2020-01-23 06:56:42 | 00:00:20
     279 |           4 |        0 |       11 |     400 |  66652 | 2020-01-23 06:57:02 | 00:00:20
(9 rows)
```

`--tps 250`

```
 slot_no | slot_leader | epoch_no | block_no | txcount |  size  |        time         |  delta   
---------+-------------+----------+----------+---------+--------+---------------------+----------
         |           1 |          |          |     259 |      0 | 2020-01-23 05:24:02 | 
     326 |           5 |        0 |        6 |     134 | 493404 | 2020-01-23 07:12:42 | 01:48:40
     327 |           3 |        0 |        7 |      34 | 124906 | 2020-01-23 07:13:02 | 00:00:20
     328 |           4 |        0 |        8 |    3229 | 537947 | 2020-01-23 07:13:22 | 00:00:20
     329 |           5 |        0 |        9 |    3229 | 537947 | 2020-01-23 07:13:42 | 00:00:20
     330 |           3 |        0 |       10 |    2742 | 456810 | 2020-01-23 07:14:02 | 00:00:20
     331 |           4 |        0 |       11 |     800 | 133288 | 2020-01-23 07:14:22 | 00:00:20
(7 rows)
```

`--tps 500`

```
  slot_no | slot_leader | epoch_no | block_no | txcount |  size  |        time         |  delta   
---------+-------------+----------+----------+---------+--------+---------------------+----------
         |           1 |          |          |     259 |      0 | 2020-01-23 05:24:02 | 
     296 |           4 |        0 |        8 |     168 | 618310 | 2020-01-23 07:02:42 | 01:38:40
     297 |           5 |        0 |        9 |    1200 | 199920 | 2020-01-23 07:03:02 | 00:00:20
     298 |           3 |        0 |       10 |    3229 | 537947 | 2020-01-23 07:03:22 | 00:00:20
     299 |           4 |        0 |       11 |    3229 | 537947 | 2020-01-23 07:03:42 | 00:00:20
     300 |           5 |        0 |       12 |    2342 | 390178 | 2020-01-23 07:04:02 | 00:00:20
(6 rows)
```

**Comment by denisshevchenko** [comment]
I repeated the experiment with these parameters:

- minimal cluster (3 core nodes)
- `50000` txs
- `100` tps
- `16` inputs
- `16` outputs
- `mempoolCap` = 4MB (still hardcoded)
- `maxTxIdsToRequest` and `maxTxToRequest` are `3` and `2`.

And blocks sizes are much bigger: rom `470KB` to `560KB`.
